rust-runtime/aws-smithy-runtime-api/src/http/headers.rs

```diff
@@ -326,13 +330,19 @@ pub use header_value::HeaderValue;
 
 type MaybeStatic = Cow<'static, str>;
 
-fn header_name(name: impl AsHeaderComponent) -> Result<http0::HeaderName, HttpError> {
+fn header_name(
+    name: impl AsHeaderComponent,
+    panic_safe: bool,
+) -> Result<http0::HeaderName, HttpError> {
     name.repr_as_http02x_header_name().or_else(|name| {
         name.into_maybe_static().and_then(|cow| {
             if cow.chars().any(|c| c.is_uppercase()) {
                 return Err(HttpError::new("Header names must be all lower case"));
             }
             match cow {
+                Cow::Borrowed(s) if panic_safe => {
+                    http0::HeaderName::try_from(s).map_err(HttpError::invalid_header_name)
+                }
```

Comment:
ideally we wouldn't need this and we'd have a cheap way of pre-validating that `from_static` would work, I guess?